whole-tale/dashboard
# Copy on Launch

### Problem
When a user launches a Tale that they do not have write access to, they are left with a half-working Tale on which they can't actually add any data or create any files. 

### Approach
When attempting to launch an instance that the user cannot access (e.g. `_accessLevel < 1`), prompt the user with a modal to copy the Tale. If the user confirms, call the new `/tale/:id/copy` endpoint before launching. This will create a copy of the existing Tale that the current user can write to, allowing them to rerun and expand the Tale of another user. We then launch this copy using the existing methods.

#### TODOs (subsequent issues)
* Don't allow user to make multiple copies
* Tie in with event notifications via `notification-stream`
* (Possibly) get rid of the modal altogether and continue without user confirmation



### How to Test
Prerequisite: at least one Tale that you do not have access to

1. Checkout and run this branch alongside [girder_wholetale's 232_tale_copy branch](https://github.com/whole-tale/girder_wholetale/pull/233), rebuild the dashboard
2. Login to the WholeTale Dashboard
3. Navigate to Browse
4. Locate a Tale that you cannot access
    * NOTE: To easily locate an inaccessible Tale, they will be missing their "delete" button
5. Click "Launch Tale" on the inaccessible Tale
    * A modal should appear prompting you to copy the Tale
6. Confirm the copy operation
    * You should see the launching Tale appear on the right side
    * You should see a status update appear beneath the launched Tale on the left side
7. Click the Tale Instance on the right side
    * You should be brought to the Run view with the new Tale instance selected
8. Navigate to Run > Files, select Tale Workspace, and Upload a File or Create a Folder
9. Once the container instance finishes starting, navigate to Run > Interact
10. Open a terminal in Jupyter or RStudio (whichever type of Tale you launched) and perform an `ls -al`
    * You should see the files/folders you uploaded in step 8 above - this confirms that you have Write Access to the new Tale Copy
